app/webdav/webdav.go

```diff
@@ -0,0 +1,407 @@
+
+	"golang.org/x/net/webdav"
+)
+
+type webdavFs struct {
+	root   *webdavFsNode
+	client *client.Client
+}
+
+func newWebDavFS(c *client.Client, br blob.Ref) (*webdavFs, error) {
+	return &webdavFs{client: c, root: &webdavFsNode{br: br, sub: make(map[string]*webdavFsNode)}}, nil
+}
+
+var (
+	_ webdav.FileSystem = (*webdavFs)(nil)
+)
+
+// fs is read only
+func (fs *webdavFs) Mkdir(ctx context.Context, name string, perm os.FileMode) error {
+	return os.ErrPermission
+}
+
+// fs is read only
+func (fs *webdavFs) RemoveAll(ctx context.Context, name string) error {
+	return os.ErrPermission
+}
+
+// fs is read only
+func (fs *webdavFs) Rename(ctx context.Context, oldName, newName string) error {
+	return os.ErrPermission
+}
+
+func (fs *webdavFs) Stat(ctx context.Context, name string) (os.FileInfo, error) {
+	f, err := fs.OpenFile(ctx, name, os.O_RDONLY, 0)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	return f.Stat()
+}
+
+func (fs *webdavFs) OpenFile(ctx context.Context, name string, flag int, perm os.FileMode) (webdav.File, error) {
+	if flag != os.O_RDONLY {
+		return nil, os.ErrPermission
+	}
+	parts := splitIntoParts(name)
+
+	n := fs.root
+	if err := fs.refreshRoot(ctx, n); err != nil {
+		return nil, fmt.Errorf("unable to refresh fs node: %w", err)
+	}
+	for i := range parts {
+		c, ok := n.sub[parts[i]]
+		if !ok {
+			return nil, os.ErrNotExist
+		}
+		if err := fs.refresh(ctx, c); err != nil {
+			return nil, fmt.Errorf("unable to refresh fs node: %w", err)
+		}
+		n = c
+	}
+	return fs.openFile(ctx, n)
+}
+
+func (fs *webdavFs) openFile(ctx context.Context, n *webdavFsNode) (*webdavFsFile, error) {
+	if n.fi.IsDir() {
+		dentries := make([]os.FileInfo, 0)
+		for _, v := range n.sub {
+			if err := fs.refresh(ctx, v); err != nil {
+				return nil, fmt.Errorf("unable to refresh fs node: %w", err)
+			}
+			dentries = append(dentries, webdavFsFileInfo{
+				isDir:   v.fi.IsDir(),
+				name:    v.fi.Name(),
+				size:    v.fi.Size(),
+				mode:    0400,
+				modTime: v.fi.ModTime(),
+			})
+		}
+		return &webdavFsFile{n: n, dentries: dentries}, nil
+	} else {
+		r, err := schema.NewFileReader(ctx, fs.client, n.br)
+		if err != nil {
+			return nil, fmt.Errorf("unable to open file to read: %w", err)
+		}
+		return &webdavFsFile{n: n, r: r}, nil
+	}
+}
+
+func splitIntoParts(name string) []string {
+	name = path.Clean(name)
+	if name == "/" {
+		return nil
+	}
+	s := make([]string, 0)
+	for {
+		dir, part := path.Split(name)
+		if part == "." || part == "/" {
+			break
+		}
+		s = append(s, part)
+		name = path.Clean(dir)
+	}
+	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
+		s[i], s[j] = s[j], s[i]
+	}
+	return s
+}
+
+type webdavFsNode struct {
+	br blob.Ref
+
+	mu  sync.Mutex
+	fi  os.FileInfo
+	sub map[string]*webdavFsNode
+
+	// cache invalidation data
+	static        bool
+	lastRefreshed time.Time
+}
+
+var refreshInterval = 1 * time.Minute
+
+func needsRefresh(n *webdavFsNode) bool {
```

Comment:
make this a method on n instead?

and if it requires its n.mu be locked, name it `needsRefreshLocked`?

Reply:
:heavy_check_mark: